File: libraries/fary/ary1_vtyp.f
      SUBROUTINE ARY1_VTYP( TYPE, VTYPE, STATUS )
*+
*  Name:
*     ARY1_VTYP

*  Purpose:
*     Validate a type specification string.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY1_VTYP( TYPE, VTYPE, STATUS )

*  Description:
*     The routine checks a data type specification string for validity.
*     To be valid it must specify one of the HDS primitive numeric data
*     types. An error is reported if the string supplied is not valid.

*  Arguments:
*     TYPE = CHARACTER * ( * ) (Given)
*        The data type specification to be checked (case insensitive).
*     VTYPE = CHARACTER * ( * ) (Returned)
*        If valid, this argument returns the data type string in upper
*        case.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Algorithm:
*     -  Check the string supplied against each permitted value in turn,
*     setting the returned argument accordingly.
*     -  If the string is not valid, then report an error.

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     9-JUN-1989  (RFWS):
*        Original version.
*     22-AUG-1989 (RFWS):
*        Corrected error in argument list of ARY1_CCPY. Also added more
*        comments.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-
      
*  Type definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_ERR'          ! ARY_ error codes

*  Arguments Given:
      CHARACTER * ( * ) TYPE

*  Arguments Returned:
      CHARACTER * ( * ) VTYPE

*  Status:
      INTEGER STATUS             ! Global status

*  External references:
      LOGICAL CHR_SIMLR          ! Case insensitive string comparison

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Check the data type string supplied against each permitted value in
*  turn, setting the returned argument accordingly.

*  ...byte data type.
      IF ( CHR_SIMLR( TYPE, '_BYTE' ) ) THEN
         CALL ARY1_CCPY( '_BYTE', VTYPE, STATUS )

*  ...unsigned byte data type.
      ELSE IF ( CHR_SIMLR( TYPE, '_UBYTE' ) ) THEN
         CALL ARY1_CCPY( '_UBYTE', VTYPE, STATUS )

*  ...double precision data type.
      ELSE IF ( CHR_SIMLR( TYPE, '_DOUBLE' ) ) THEN
         CALL ARY1_CCPY( '_DOUBLE', VTYPE, STATUS )

*  ...integer data type.
      ELSE IF ( CHR_SIMLR( TYPE, '_INTEGER' ) ) THEN
         CALL ARY1_CCPY( '_INTEGER', VTYPE, STATUS )

*  ...real data type.
      ELSE IF ( CHR_SIMLR( TYPE, '_REAL' ) ) THEN
         CALL ARY1_CCPY( '_REAL', VTYPE, STATUS )

*  ...word data type.
      ELSE IF ( CHR_SIMLR( TYPE, '_WORD' ) ) THEN
         CALL ARY1_CCPY( '_WORD', VTYPE, STATUS )

*  ...unsigned word data type.
      ELSE IF ( CHR_SIMLR( TYPE, '_UWORD' ) ) THEN
         CALL ARY1_CCPY( '_UWORD', VTYPE, STATUS )

*  If the data type string is invalid, then report an error.
      ELSE
         STATUS = ARY__TYPIN
         CALL MSG_SETC( 'BADTYPE', TYPE )
         CALL ERR_REP( 'ARY1_VTYP_BAD',
     :   'Invalid array data type ''^BADTYPE'' specified (possible ' //
     :   'programming error).', STATUS )
      END IF
       
*  Call error tracing routine and exit.
      IF ( STATUS .NE. SAI__OK ) CALL ARY1_TRACE( 'ARY1_VTYP', STATUS )

      END

File: libraries/ary/ary1Vtyp.c
#include "sae_par.h"
#include "ary1.h"
#include "mers.h"
#include "ary_err.h"
#include "dat_par.h"
#include <string.h>

void ary1Vtyp( const char *type, char vtype[DAT__SZTYP + 1], int *status ) {
/*
*+
*  Name:
*     ary1Vtyp

*  Purpose:
*     Validate a type specification string.

*  Synopsis:
*     void ary1Vtyp( const char *type, char vtype[DAT__SZTYP + 1], int *status )

*  Description:
*     This function checks a data type specification string for validity.
*     To be valid it must specify one of the HDS primitive numeric data
*     types. An error is reported if the string supplied is not valid.

*  Parameters:
*     type
*        The data type specification to be checked (case insensitive).
*     vtype
*        If valid, this argument returns the data type string in upper
*        case.
*     status
*        The global status.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     03-JUL-2017 (DSB):
*        Original version, based on equivalent Fortran routine by RFWS.

*-
*/

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Check the data type string supplied against each permitted value in
   turn, setting the returned argument accordingly. */

/* ...byte data type. */
   if( !strcasecmp( type, "_BYTE" ) ){
      ary1Ccpy( "_BYTE", DAT__SZTYP + 1, vtype, status );

/* ...unsigned byte data type. */
   } else if( !strcasecmp( type, "_UBYTE" ) ){
      ary1Ccpy( "_UBYTE", DAT__SZTYP + 1, vtype, status );

/* ...double precision data type. */
   } else if( !strcasecmp( type, "_DOUBLE" ) ){
      ary1Ccpy( "_DOUBLE", DAT__SZTYP + 1, vtype, status );

/* ...integer data type. */
   } else if( !strcasecmp( type, "_INTEGER" ) ){
      ary1Ccpy( "_INTEGER", DAT__SZTYP + 1, vtype, status );

/* ...real data type. */
   } else if( !strcasecmp( type, "_REAL" ) ){
      ary1Ccpy( "_REAL", DAT__SZTYP + 1, vtype, status );

/* ...word data type. */
   } else if( !strcasecmp( type, "_WORD" ) ){
      ary1Ccpy( "_WORD", DAT__SZTYP + 1, vtype, status );

/* ...unsigned word data type. */
   } else if( !strcasecmp( type, "_UWORD" ) ){
      ary1Ccpy( "_UWORD", DAT__SZTYP + 1, vtype, status );

/* ...64-bit integer data type. */
   } else if( !strcasecmp( type, "_INT64" ) ){
      ary1Ccpy( "_INT64", DAT__SZTYP + 1, vtype, status );

/* If the data type string is invalid, then report an error. */
   } else {
      *status = ARY__TYPIN;
      msgSetc( "BADTYPE", type );
      errRep( " ", "Invalid array data type '^BADTYPE' specified (possible "
              "programming error).", status );
   }

/* Call error tracing routine and exit. */
   if( *status != SAI__OK ) ary1Trace( "ary1Vtyp", status );

}
